theia-broker/broker.go

```diff
@@ -0,0 +1,216 @@
+
+	broker.broker.PrintPlan(metas)
+
+	broker.broker.PrintInfo("Starting plugins processing")
+	for _, meta := range metas {
+		err := broker.processPlugin(meta)
+		if err != nil {
+			broker.broker.PubFailed(err.Error())
+			broker.broker.PrintFatal(err.Error())
+		}
+	}
+
+	if ok, status := storage.SetStatus(model.StatusDone); !ok {
+		err := fmt.Sprintf("Setting '%s' broker status failed. Broker has '%s' state", model.StatusDone, status)
+		broker.broker.PubFailed(err)
+		broker.broker.PrintFatal(err)
+	}
+
+	tooling, err := storage.Tooling()
+	if err != nil {
+		broker.broker.PubFailed(err.Error())
+		broker.broker.PrintFatal(err.Error())
+	}
+	toolingBytes, err := json.Marshal(tooling)
+	if err != nil {
+		broker.broker.PubFailed(err.Error())
+		broker.broker.PrintFatal(err.Error())
+	}
+
+	broker.broker.PrintInfo("All plugins have been successfully processed")
+	broker.broker.PubDone(string(toolingBytes))
+	broker.broker.CloseConsumers()
+}
+
+// PushEvents sets given tunnel as consumer of broker events.
+func (broker *TheiaPluginBroker) PushEvents(tun *jsonrpc.Tunnel) {
+	broker.broker.PushEvents(tun, model.BrokerStatusEventType, model.BrokerResultEventType, model.BrokerLogEventType)
+}
+
+func (broker *TheiaPluginBroker) processPlugin(meta model.PluginMeta) error {
+	broker.broker.PrintDebug("Stared processing plugin '%s:%s'", meta.ID, meta.Version)
+	url := meta.URL
+
+	workDir, err := ioutil.TempDir("", "che-plugin-broker")
+	if err != nil {
+		return err
+	}
+
+	archivePath := filepath.Join(workDir, "testArchive.tar.gz")
+	pluginPath := filepath.Join(workDir, "testArchive")
+
+	// Download an archive
+	broker.broker.PrintDebug("Downloading archive '%s' for plugin '%s:%s' to '%s'", url, meta.ID, meta.Version, archivePath)
+	err = files.Download(url, archivePath)
+	if err != nil {
+		return err
+	}
+
+	// Untar it
+	broker.broker.PrintDebug("Untarring archive '%s' for plugin '%s:%s' to '%s'", url, meta.ID, meta.Version, archivePath)
+	err = files.Untar(archivePath, pluginPath)
+	if err != nil {
+		return err
+	}
+
+	broker.broker.PrintDebug("Resolving Che plugins for '%s:%s'", meta.ID, meta.Version)
+	err = broker.resolveToolingConfig(pluginPath)
+	if err != nil {
+		return err
+	}
+
+	broker.broker.PrintDebug("Copying dependencies for '%s:%s'", meta.ID, meta.Version)
+	return broker.copyDependencies(pluginPath)
+}
+
+func (broker *TheiaPluginBroker) resolveToolingConfig(workDir string) error {
+	toolingConfPath := filepath.Join(workDir, "che-plugin.yaml")
+	f, err := ioutil.ReadFile(toolingConfPath)
+	if err != nil {
+		return err
+	}
+
+	tooling := &model.ToolingConf{}
+	if err := yaml.Unmarshal(f, tooling); err != nil {
+		return err
+	}
+
+	broker.addPortToTooling(tooling)
+
+	return storage.AddTooling(tooling)
+}
+
+func (broker *TheiaPluginBroker) copyDependencies(workDir string) error {
+	depsConfPath := filepath.Join(workDir, "che-dependency.yaml")
+	if _, err := os.Stat(depsConfPath); os.IsNotExist(err) {
+		return nil
```

Comment:
maybe we should return some meaningful error?

Reply:
No, it is a valid case